### PR TITLE
add support for generating swap statements for dependencies (REVIEW)

### PR DESCRIPTION
**Pull request by boegel**
required for stable Cray support (cfr. https://github.com/hpcugent/easybuild-framework/issues/1390)

### CODE REVIEW

**Comment by boegel** [comment]
@wpoely86: still needs tests, but can you give this an early review?

@gppezzi: this is required for the `CrayToolchain` easyblock

**Comment by boegel** [comment]
cc @pforai

File: easybuild/framework/easyblock.py [hunk 1]
```diff
@@ -865,12 +865,13 @@ def make_devel_module(self, create_in_builddir=False):
         # cleanup: unload fake module, remove fake module dir
         self.clean_up_fake_module(fake_mod_data)
 
-    def make_module_dep(self):
+    def make_module_dep(self, swap_info=None):
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: document arguments

File: easybuild/framework/easyblock.py [hunk 2]
```diff
@@ -865,12 +865,13 @@ def make_devel_module(self, create_in_builddir=False):
...
         """
         Make the dependencies for the module file.
         """
         deps = []
         mns = ActiveMNS()
+        swap_info = swap_info or {}
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: Can you just not as well make the default arg `{}`?

Comment (boegel) [inline easybuild/framework/easyblock.py]: no, mutable values as default value are of the devil, see http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

File: easybuild/tools/module_generator.py [hunk 3]
```diff
@@ -128,6 +128,25 @@ def conditional_statement(self, condition, body, negative=False):
...
+        raise NotImplementedError
+
+    def swap_module(self, mod_names, recursive_unload=False):
+        """Generate swap statement for specified modules."""
+        # swap via unload/load
+        # 'swap' command is not used, because:
+        # i) it would fail if the module being swapped out is not loaded
+        # ii) it would do weird stuff when the module in which it is used in gets unloaded (?)
+        comment = self.comment("'safe' swap: unload %s when loaded, then load %s" % mod_names).strip()
+        unload = self.unload_module(mod_names[0]).strip()
```
Comment (wpoely86) [inline easybuild/tools/module_generator.py]: add check if `mod_names` is a tuple?

File: easybuild/tools/module_generator.py [hunk 4]
```diff
@@ -128,6 +128,25 @@ def conditional_statement(self, condition, body, negative=False):
...
 
+    def load_module(self, mod_name, recursive_unload=False):
+        """Generate load statement for specified module."""
+        raise NotImplementedError
+
+    def unload_module(self, mod_name, conditional=True):
+        """Generate unload statement for specified module."""
+        raise NotImplementedError
+
+    def swap_module(self, mod_names, recursive_unload=False):
```
Comment (wpoely86) [inline easybuild/tools/module_generator.py]: doc arguments

Comment (wpoely86) [inline easybuild/tools/module_generator.py]: ( everywhere )

File: easybuild/tools/module_generator.py (hunk 3 above)
Comment (boegel) [inline easybuild/tools/module_generator.py]: nah, that's like an assert in the code, we don't do that anywhere?

File: easybuild/framework/easyblock.py (hunk 1 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: done

**Comment by boegel** [comment]
Going in, thanks for the review @wpoely86!

**Pull request merged**